File: design_patterns/C++/Structural/flyweight.cxx
#include <iostream>
#include <map>

// Anything that will be cached is flyweight.
// Types of tea here will be flyweights.
class KarakTea {};

// Acts as a factory and saves the tea
class TeaMaker {
  protected:
	std::map<std::string, KarakTea*> availableTea;

  public:
	KarakTea* make(std::string preference) {
		if (!this->availableTea[preference]) {
			this->availableTea[preference] = new KarakTea();
		}
		return this->availableTea[preference];
	}
};

class TeaShop {
  protected:
	std::map<int, KarakTea*> orders;
	TeaMaker* teaMaker = nullptr;

  public:
	TeaShop(TeaMaker* teaMaker) {
		this->teaMaker = teaMaker;
	}

	void takeOrder(std::string teaType, int table) {
		this->orders[table] = this->teaMaker->make(teaType);
	}

	void serve() {
		for (auto order : this->orders) {
			std::cout << "Serving tea to table #" << order.first << std::endl;
		}
	}
};

int main() {
	auto teaMaker = new TeaMaker();
	auto shop = new TeaShop(teaMaker);
	shop->takeOrder("less sugar", 1);
	shop->takeOrder("more milk", 2);
	shop->takeOrder("without sugar", 5);
	shop->serve();
	// =======================
	// Serving tea to table #1
	// Serving tea to table #2
	// Serving tea to table #5
	// =======================
	return 0;
}

File: design_patterns/Python/Structural/flyweight.py
from collections import defaultdict


class KarakTea:
    # Anything that will be cached is flyweight
    # Types of tea here will be flyweights
    pass


class TeaMaker:
    # Acts as a factory and saves the tea

    def __init__(self):
        self.availible_tea = defaultdict(list)

    def make(self, preference):
        if len(self.availible_tea[preference]) == 0:
            self.availible_tea[preference] = KarakTea()
        return self.availible_tea[preference]


class TeaShop:

    def __init__(self, tea_maker):
        self.tea_maker = tea_maker
        self.orders = dict()

    def take_order(self, tea_type, table):
        self.orders[table] = self.tea_maker.make(tea_type)

    def serve(self):
        for table in self.orders.keys():
            print('Serving tea to table #' + str(table))


if __name__ == '__main__':

    tea_maker = TeaMaker()
    shop = TeaShop(tea_maker)
    shop.take_order('less suger', 1)
    shop.take_order('more milk', 2)
    shop.take_order('without suger', 5)
    shop.serve()
    # Serving tea to table #1
    # Serving tea to table #2
    # Serving tea to table #5
